indexer/services/roundtable/src/tasks/uncross-orderbook.ts

```diff
@@ -0,0 +1,119 @@
+import { logger } from '@dydxprotocol-indexer/base';
+import { OrderSide, PerpetualMarketFromDatabase, PerpetualMarketTable } from '@dydxprotocol-indexer/postgres';
+import { OrderbookLevels, OrderbookLevelsCache } from '@dydxprotocol-indexer/redis';
+import Big from 'big.js';
+
+import { redisClient } from '../helpers/redis';
+
+/**
+ * Task to uncross the orderbook.
+ */
+export default async function runTask(): Promise<void> {
+  const markets: PerpetualMarketFromDatabase[] = await PerpetualMarketTable.findAll({}, []);
+
+  for (let i: number = 0; i < markets.length; i++) {
+    const market: PerpetualMarketFromDatabase = markets[i];
+
+    const crossedOrderbookLevels: OrderbookLevels = await OrderbookLevelsCache.getOrderBookLevels(
+      market.ticker,
+      redisClient,
+      {
+        removeZeros: true,
+        sortSides: true,
+        uncrossBook: false,
+      },
+    );
+
+    // Check if the orderbook is crossed
+    if (isOrderbookCrossed(crossedOrderbookLevels)) {
+      await uncrossOrderbook(market, crossedOrderbookLevels);
+    }
+  }
+}
+
+/**
+ * Check if the orderbook is crossed.
+ */
+function isOrderbookCrossed(orderbookLevels: OrderbookLevels): boolean {
+  if (orderbookLevels.bids.length > 0 && orderbookLevels.asks.length > 0) {
+    const bestBid = Big(orderbookLevels.bids[0].humanPrice);
+    const bestAsk = Big(orderbookLevels.asks[0].humanPrice);
+    return bestBid.gte(bestAsk);
+  }
+  return false;
+}
+
+/**
+ * Uncross the orderbook by removing overlapping bid and ask levels.
+ * Remove the price levels that are older.
+ */
+async function uncrossOrderbook(
+  market: PerpetualMarketFromDatabase,
+  orderbookLevels: OrderbookLevels,
+): Promise<void> {
+  const ticker = market.ticker;
+
+  // Remove overlapping levels
+  let ai = 0;
+  let bi = 0;
+
+  while (
+    ai < orderbookLevels.asks.length &&
+    bi < orderbookLevels.bids.length &&
+    Big(orderbookLevels.bids[bi].humanPrice).gte(Big(orderbookLevels.asks[ai].humanPrice))
+  ) {
+    // Remove the older side
+    if (
+      Number(orderbookLevels.bids[bi].lastUpdated) > Number(orderbookLevels.asks[ai].lastUpdated)
+    ) {
+      ai += 1;
+    } else {
+      bi += 1;
+    }
+  }
+
+  // Remove crossed levels from Redis
+  const removeBidLevels = orderbookLevels.bids.slice(0, bi);
+  const removeAskLevels = orderbookLevels.asks.slice(0, ai);
+
+  logger.info({
+    at: 'uncrossOrderbook#uncrossOrderbook',
+    message: `Uncrossing orderbook for ${ticker}`,
+    removedBids: JSON.stringify(removeBidLevels),
+    removedAsks: JSON.stringify(removeAskLevels),
+  });
+
+  for (const bid of removeBidLevels) {
```

Comment:
nit: Add metrics for the number of bids / asks deleted per iteration and number of crossed orderbook?